### PR TITLE
Server synchronisation

### PR DESCRIPTION
**Pull request by raphaelMrci**
## Description
Add server synchronisation

## Lien tache Jira
https://leomaman.atlassian.net/browse/TYPE-118?atlOrigin=eyJpIjoiOWYxYmFmMTJjMjcwNGY0Njk1ZmRhZjhkNmY0NjA2NjMiLCJwIjoiaiJ9

https://leomaman.atlassian.net/browse/TYPE-119?atlOrigin=eyJpIjoiMDcwNmM1M2NlYmRjNDRjOWFjY2I2NDY2ZmFjZjgwY2EiLCJwIjoiaiJ9

## Changement
Authorite server

Synchronisation des mouvements joueurs

### Summary by CodeRabbit

- New Feature: Added support for player position tracking and handling in the game, enhancing real-time gameplay experience.
- Refactor: Updated the codebase to use Boost.Asio library separately, improving the modularity of the code.
- New Feature: Introduced a new `ButtonEntity` class for better UI interaction.
- Refactor: Modified various components and systems to remove the `PacketManager` parameter, simplifying the function signatures and improving code readability.
- New Feature: Added a new function `getEntityByID` to the `ECS::Scene` class, providing a way to fetch entities by their ID.
- Refactor: Updated the server's main loop to run in a separate thread, improving the server's performance and responsiveness.

### CODE REVIEW

File: Network/Packet.hpp [hunk 1]
```diff
@@ -14,10 +14,14 @@
 #include "MoveData.hpp"
 #include "PacketType.hpp"
 #include "StartData.hpp"
+#include "PlayersPos.hpp"
 
 /**
  * @brief Network protocol
  */
+
+#define MILLISECONDS_TICKS 20
```
Comment (christophechr) [inline Network/Packet.hpp:23]: The name isn't explicit

File: Network/PacketManager.hpp [hunk 2]
```diff
@@ -56,7 +59,7 @@ namespace Network {
              * @see Network::PacketType
              * @see Network::data
              */
-            std::unique_ptr<Packet> createPacket(
+            static std::unique_ptr<Packet> createPacket(
```
Comment (christophechr) [inline Network/PacketManager.hpp:62]: Instead of using static public method, declare it as private method if it's the objective

**Review by christophechr — CHANGES_REQUESTED** [review]

File: Network/PacketManager.cpp [hunk 3]
```diff
@@ -7,8 +7,9 @@
 
 #include <cstring>
 #include <iostream>
-#include <string>
+#include <utility>
 #include <vector>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (mangasteak) [reply Network/PacketManager.cpp:12]: I AGREE

File: Network/PacketManager.cpp [hunk 4]
```diff
@@ -32,17 +33,17 @@
     std::unique_ptr<Network::Packet> packet =
     std::make_unique<Network::Packet>();
 
+    std::cout << "bytes_size: " << bytes_size << std::endl;
```
[the remark this answers is not in the document]
Comment (mangasteak) [reply Network/PacketManager.cpp]: should probably be removed

File: Network/PacketManager.cpp [hunk 5]
```diff
@@ -62,27 +63,39 @@
...
 {
     if (_handlers.find(packet.type) == _handlers.end()) {
         std::cerr << "No handler for packet type: "
                   << static_cast<int>(packet.type) << std::endl;
         return;
     }
-    std::cout << "Handling packet type: " << static_cast<int>(packet.type)
-              << std::endl;
-    _handlers[packet.type](packet);
+    std::cout << "Handling packet" << std::endl;
```
[the remark this answers is not in the document]
Comment (mangasteak) [reply Network/PacketManager.cpp]: here also

File: client/ecs/components/ClickComponent.cpp [hunk 6]
```diff
@@ -6,9 +6,10 @@
 */
 
 #include <iostream>
+#include <utility>
 #include "ClickComponent.hpp"
 
-ECS::ClickComponent::ClickComponent(sf::Rect<int> rect, ECS::eventCallback callback, sf::RenderWindow &window) : ECS::EventComponent(callback), _rect(rect), _window(window)
+ECS::ClickComponent::ClickComponent(sf::Rect<int> rect, ECS::eventCallback callback, sf::RenderWindow &window) : ECS::EventComponent(std::move(callback)), _rect(rect), _window(window)
```
[the remark this answers is not in the document]
Comment (mangasteak) [reply client/ecs/components/ClickComponent.cpp:12]: +1

File: client/ecs/models/ButtonEntity.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    sf::Rect<int> rect;
+    rect.left = 0;
+    rect.top = 0;
+    rect.width = texture.getSize().x;
+    rect.height = texture.getSize().y;
+
+    addComponent(std::make_shared<ECS::SpriteComponent>(texture, rect));
+    addComponent(std::make_shared<ECS::ScaleComponent>(0.5f, 0.5f));
+}
```
[the remark this answers is not in the document]
Comment (mangasteak) [reply client/ecs/models/ButtonEntity.cpp:33]: a voir

**Review by mangasteak — COMMENTED** [review]
a voir pour certain points

File: server/main.cpp [hunk 8]
```diff
@@ -5,17 +5,21 @@
...
 #include "Packet.hpp"
 #include "Server.hpp"
+#include "ServerCore.hpp"
 
-int main(void)
+int main()
 {
     boost::asio::io_service io_service;
+    RType::Server server(io_service, 4242);
+    ECS::ServerCore serverCore(server);
```
[the remark this answers is not in the document]
Comment (raphaelMrci) [reply server/main.cpp:17]: ok boomer

File: server/ServerCore.cpp [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+        lastFrameTime = std::chrono::high_resolution_clock::now();
+
+        _server.packetManager.executeRecvPacketsQueue();
+        for (auto &system : _systems) {
+            system->update(
+            sceneManager, _deltaTime, _server.packetManager.sendPacketsQueue);
+        }
+        _server.sendPackets();
+        std::this_thread::sleep_for(std::chrono::milliseconds(MILLISECONDS_TICKS));
+    }
```
[the remark this answers is not in the document]
Comment (raphaelMrci) [reply server/ServerCore.cpp:72]: ok boomer

File: server/ServerCore.hpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            void _initEntities();
+
+            void _initHandlers(Network::PacketManager &packetManager);
+
+            void _handlerStartGame(Network::Packet &packet, const udp::endpoint &endpoint);
+            
+            void _handlerMoveUp(Network::Packet &packet, const udp::endpoint &endpoint);
+            void _handlerMoveDown(Network::Packet &packet, const udp::endpoint &endpoint);
+            void _handlerMoveLeft(Network::Packet &packet, const udp::endpoint &endpoint);
+            void _handlerMoveRight(Network::Packet &packet, const udp::endpoint &endpoint);
```
[the remark this answers is not in the document]
Comment (raphaelMrci) [reply server/ServerCore.hpp:46]: What do you mean ?

File: Network/PacketManager.hpp (hunk 2 above)
Comment (raphaelMrci) [reply Network/PacketManager.hpp:62]: Nop, ca permet de ne pas la copier a chaque creation de classe et y acceder sans avoir besoin d'acceder a une class built

**Review by mangasteak — APPROVED** [review]

**Review by christophechr — APPROVED** [review]

**Pull request merged**